File: reverse-vowels-of-a-string/solution345.go
// https://leetcode.com/problems/reverse-vowels-of-a-string/

package main

import "fmt"

var vowels = map[rune]bool{
	'a': true, 'e': true, 'i': true, 'o': true, 'u': true,
	'A': true, 'E': true, 'I': true, 'O': true, 'U': true,
}

// T: O(n), S: O(n)
func reverseVowels(s string) string {
	arr := []rune(s)
	left, right := 0, len(s)-1
	for left < right {
		if !vowels[arr[left]] {
			left++
		} else if !vowels[arr[right]] {
			right--
		} else {
			arr[left], arr[right] = arr[right], arr[left]
			left++
			right--
		}
	}
	return string(arr)
}

func main() {
	s1 := "hello"
	fmt.Println(reverseVowels(s1)) // "holle"
	s2 := "leetcode"
	fmt.Println(reverseVowels(s2)) // "leotcede"
}

File: reverse-vowels-of-a-string/Solution345.java
// https://leetcode.com/problems/reverse-vowels-of-a-string/

import java.util.Set;

class Solution345 {

    private static final Set<Character> VOWELS = Set.of(
            'A', 'E', 'I', 'O', 'U',
            'a', 'e', 'i', 'o', 'u');

    // T: O(n), S: O(n)
    public String reverseVowels(String s) {
        char[] arr = s.toCharArray();
        int len = arr.length, left = 0, right = len - 1;
        while (left < right) {
            if (!VOWELS.contains(arr[left])) {
                left++;
            } else if (!VOWELS.contains(arr[right])) {
                right--;
            } else {
                swap(arr, left++, right--);
            }
        }
        return new String(arr);
    }

    private static void swap(char[] arr, int i, int j) {
        if (i != j) {
            char temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    public static void main(String[] args) {
        var sol = new Solution345();
        var s1 = "hello";
        System.out.println(sol.reverseVowels(s1)); // "holle"
        var s2 = "leetcode";
        System.out.println(sol.reverseVowels(s2)); // "leotcede"
    }

}
